File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> output(n, 1);
        int left = 1, right = 1;

        for(int i = 0;i < n; i++) {
            output[i] = output[i] * left;
            left = left*nums[i];
        }

        for(int i = n-1; i >= 0 ; i--) {
            output[i] = output[i] * right;
            right = right * nums[i];
        }

        return output;
    }
};


// 1 2 4 6
// 1 1 2 8
// 48  24  6  1 

// 1 2 6 1
// = 48 24 12 8

File: Data Structures & Algorithms/kth-largest-element-in-an-array/submission-2.java
class Solution {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        int output;

        for(int i = 0;i < nums.length; i++) {
            if(minHeap.size() < k) {
                minHeap.add(nums[i]);
            }
            else if(minHeap.peek() < nums[i]) {
                minHeap.remove();
                minHeap.add(nums[i]);
            }
        }

        output = minHeap.poll();

        return output;
    }
}
